File: java/Descifrado.java
import java.nio.file.Path;
import java.util.HexFormat;

public final class Descifrado {
    public static void main(String[] args) throws Exception {
        if (args.length != 9) {
            System.err.println("Uso: java Descifrado <x_r-1_u16.bin> <x_r_u16.bin> <out.png|bmp|raw> <ancho> <alto> <canales> <rondas> <Z_hex_64> <salt_hex_64>");
            System.exit(1);
        }

        String prevPath = args[0];
        String curPath = args[1];
        String outPath = args[2];
        int width = Integer.parseInt(args[3]);
        int height = Integer.parseInt(args[4]);
        int channels = Integer.parseInt(args[5]);
        int rounds = Integer.parseInt(args[6]);
        byte[] z = HexFormat.of().parseHex(args[7]);
        byte[] salt = HexFormat.of().parseHex(args[8]);

        short[] prev = Common.readU16(Path.of(prevPath));
        short[] cur = Common.readU16(Path.of(curPath));
        int elems = width * height * channels;
        if (prev.length != elems || cur.length != elems) {
            throw new IllegalArgumentException("Tamano invalido de archivos cifrados.");
        }

        Common.RcaContext ctx = new Common.RcaContext(height, width, channels, rounds);
        var session = Llaves.deriveSession(z, salt, ctx);

        short[] nextState = cur;
        short[] curState = prev;
        short[] curPerm = new short[elems];
        short[] prevRec = new short[elems];

        for (int t = rounds - 1; t >= 0; t--) {
            var rk = Llaves.deriveRound(session, ctx, t);
            System.arraycopy(curState, 0, curPerm, 0, elems);
            Permutaciones.applyU16(curPerm, height, width, channels, rk.permIndex());
            int[][] k1 = Automata.kernelFromMoore8(rk.moore1());
            int[][] k2 = Automata.kernelFromMoore8(rk.moore2());
            Automata.stepU16(nextState, curPerm, prevRec, height, width, channels, k1, k2, t, Llaves.boundaryFromRound(rk));
            short[] tmp = nextState; nextState = curState; curState = prevRec; prevRec = tmp;
        }

        byte[] output = new byte[elems];
        for (int i = 0; i < elems; i++) {
            int v = Short.toUnsignedInt(nextState[i]);
            output[i] = (byte)(v == 0 ? 255 : v - 1);
        }
        Common.savePreview(output, width, height, channels, outPath);

        System.out.println("Descifrado completado.");
        System.out.println("Entrada estado previo: " + prevPath);
        System.out.println("Entrada estado final: " + curPath);
        System.out.println("Salida imagen: " + outPath);
        System.out.printf("Dimensiones: %dx%d x %d, rondas=%d%n", width, height, channels, rounds);
    }
}

File: cs/Descifrado.cs
using System;
using System.IO;
using TTCrypto;

internal static class Descifrado
{
    private static int Main(string[] args)
    {
        try
        {
            if (args.Length != 9)
            {
                Console.Error.WriteLine("Uso: Descifrado <x_r-1_u16.bin> <x_r_u16.bin> <out.png|bmp|tif|tiff|raw> <ancho> <alto> <canales> <rondas> <Z_hex_64> <salt_hex_64>");
                return 1;
            }

            string prevPath = args[0];
            string curPath = args[1];
            string outPath = args[2];
            uint width = uint.Parse(args[3]);
            uint height = uint.Parse(args[4]);
            ushort channels = ushort.Parse(args[5]);
            ushort rounds = ushort.Parse(args[6]);
            byte[] z = HexUtil.FromHex(args[7]);
            byte[] salt = HexUtil.FromHex(args[8]);

            var prev = ReadU16(prevPath);
            var cur = ReadU16(curPath);
            var elems = checked((int)(width * height * channels));
            if (prev.Length != elems || cur.Length != elems)
                throw new InvalidDataException("Tamano invalido de los archivos cifrados.");

            var ctx = new RcaContext(height, width, channels, rounds);
            var session = Keys.DeriveSession(z, salt, ctx);

            var nextState = cur;
            var curState = prev;
            var curPerm = new ushort[elems];
            var prevRec = new ushort[elems];

            for (int t = rounds - 1; t >= 0; t--)
            {
                var rk = Keys.DeriveRound(session, ctx, (uint)t);
                Array.Copy(curState, curPerm, elems);
                Permutations.ApplyU16(curPerm, height, width, channels, rk.PermIndex);
                var k1 = Automata.KernelFromMoore8(rk.Moore1);
                var k2 = Automata.KernelFromMoore8(rk.Moore2);
                Automata.StepU16(nextState, curPerm, prevRec, height, width, channels, k1, k2, (uint)t, BoundaryHelper.FromRound(rk));
                (nextState, curState, prevRec) = (curState, prevRec, nextState);
            }

            var output = new byte[elems];
            for (var i = 0; i < elems; i++)
            {
                var v = nextState[i];
                output[i] = v == 0 ? (byte)255 : (byte)(v - 1);
            }

            ImageUtil.SavePreview(output, width, height, channels, outPath);
            Console.WriteLine("Descifrado completado.");
            Console.WriteLine($"Entrada estado previo: {prevPath}");
            Console.WriteLine($"Entrada estado final: {curPath}");
            Console.WriteLine($"Salida imagen: {outPath}");
            Console.WriteLine($"Dimensiones: {width}x{height} x {channels}, rondas={rounds}");
            return 0;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex.Message);
            return 2;
        }
    }

    private static ushort[] ReadU16(string path)
    {
        var bytes = File.ReadAllBytes(path);
        if ((bytes.Length & 1) != 0) throw new InvalidDataException("Archivo u16 invalido.");
        var data = new ushort[bytes.Length / 2];
        Buffer.BlockCopy(bytes, 0, data, 0, bytes.Length);
        return data;
    }
}
